### PR TITLE
Point in time recovery

### PR DESCRIPTION
**Pull request by Horusiath**
This PR introduces a new command:

```
turso db restore $DATABASE --timestamp 2023-09-25T14:53:00 --yes
```

Which is going to trigger database point-in-time recovery procedure. It's going to call turso API: `/v1/databases/:database/restore`.

### CODE REVIEW

File: internal/turso/turso.go [hunk 1]
```diff
@@ -166,3 +166,7 @@ func (t *Client) Upload(path string, fileData *os.File) (*http.Response, error)
 func (t *Client) Delete(path string, body io.Reader) (*http.Response, error) {
 	return t.do("DELETE", path, body)
 }
+
+func (t *Client) Restore(path string, body io.Reader) (*http.Response, error) {
+	return t.do("DELETE", path, body)
```
Comment (haaawk) [inline internal/turso/turso.go:171]: Is restore a HTTP DELETE?

File: internal/cmd/timestamp_flag.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package cmd
+
+import "github.com/spf13/cobra"
+
+var timestampFlag string
+
+func addDbTimestampFlag(cmd *cobra.Command) {
+	restoreCmd.Flags().StringVarP(&timestampFlag, "timestamp", "t", "", "UTC timestamp for the version of database at a given point in time.")
```
Comment (haaawk) [inline internal/cmd/timestamp_flag.go:8]: What's the format of the timestamp? Just int? In what unit?

Comment (haaawk) [reply internal/cmd/timestamp_flag.go:8]: Ok. I can see it here https://github.com/tursodatabase/turso-cli/pull/637/files#diff-e963bae42bbc05e1ecbaa07a34c3a585f0278493311e24f2a4ab030e8267c940R19

File: internal/cmd/db_restore.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+func init() {
+	dbCmd.AddCommand(restoreCmd)
+	restoreCmd.Flags().BoolVarP(&yesFlag, "yes", "y", false, "Confirms the restoration of all locations of the database.")
+	addDbTimestampFlag(restoreCmd)
+	restoreCmd.RegisterFlagCompletionFunc("instance", completeInstanceName)
+}
+
+var restoreCmd = &cobra.Command{
+	Use:               "restore database_name --timestamp 2023-09-25T14:53:00",
```
Comment (haaawk) [inline internal/cmd/db_restore.go:19]: I think timestamp should not be a flag but an argument. Without timestamp this operation makes no sense. You're also not checking if timestamp is set in the RunE. Which just proves my point.

File: internal/cmd/db_restore.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	"github.com/chiselstrike/iku-turso-cli/internal"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	dbCmd.AddCommand(restoreCmd)
+	restoreCmd.Flags().BoolVarP(&yesFlag, "yes", "y", false, "Confirms the restoration of all locations of the database.")
+	addDbTimestampFlag(restoreCmd)
+	restoreCmd.RegisterFlagCompletionFunc("instance", completeInstanceName)
```
Comment (haaawk) [inline internal/cmd/db_restore.go:15]: This should probably go away, no?

File: internal/cmd/utils.go [hunk 5]
```diff
@@ -131,6 +131,22 @@ func printTable(header []string, data [][]string) {
...
+	defer s.Stop()
+
+	if err := client.Databases.Restore(name, timestamp); err != nil {
+		return err
+	}
+	s.Stop()
+	elapsed := time.Since(start)
+
+	fmt.Printf("Restored database %s in %d seconds.\n", internal.Emph(name), int(elapsed.Seconds()))
+	invalidateDatabasesCache()
```
Comment (haaawk) [inline internal/cmd/utils.go:146]: We don't need this, do we?

File: internal/cmd/db_restore.go (hunk 3 above)
Comment (Horusiath) [reply internal/cmd/db_restore.go:19]: General idea was that in the future we might want to do restoration based on different kinds of parameters (sqld natively supports timestamp-based or generation-based restoration).

File: internal/cmd/timestamp_flag.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package cmd
+
+import "github.com/spf13/cobra"
+
+var timestampFlag string
+
+func addDbTimestampFlag(cmd *cobra.Command) {
```
Comment (haaawk) [inline internal/cmd/timestamp_flag.go:7]: Is this used anywhere?

**Review by haaawk — APPROVED** [review]

**Comment by Horusiath** [comment]
After yesterday's talks and discussion with @athoscouto it looks like it will be more reasonable to change that feature into optional param as part of `turso db create {forked-db-name} --from-db {parent-db-name} --timestamp {timestamp}`.

**Pull request closed without merging**